### PR TITLE
Don't fail silently when patching files

### PR DESCRIPTION
**Pull request by tru**
After searching for why my files where not patched correctly
way to long I came to the conclusion that the patch tool actually
silently failed. This is pretty bad behavior if something changes
so I have added a exception for this.

### CODE REVIEW

File: conans/tools.py [hunk 1]
```diff
@@ -134,4 +133,6 @@ def patch(base_path=None, patch_file=None, patch_string=None):
     else:
         patchset = fromstring(patch_string.encode())
 
-    patchset.apply(root=base_path)
+    if not patchset.apply(root=base_path):
+        raise Exception("Failed to apply patch: %s" % patch_file)
```
Comment (memsharded) [inline conans/tools.py]: Would you mind changing it to a ConanException?
ConanExceptions are more controlled, produced a cleaner output (without a traceback that is ugly for most users), as we know it is something that conan produced. On the contrary regular Exceptions are non foreseen problems, bugs, and other issues, and even if we capture them (eg. in ``Conanfile.config()`` method, to allow the package creators to raise regular Exceptions) in some contexts, the usual is that they produce an unformatted stack trace dump, that by design is considered always a conan bug.
Makes sense this design criteria?

Comment (tru) [inline conans/tools.py]: Yeah makes sense. I can update that for sure.

**Comment by tru** [comment]
@memsharded Updated to use `ConanException` instead.

**Comment by memsharded** [comment]
Excellent, thank you!

**Pull request merged**